presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/PushTopNThroughOuterJoinRuleSet.java

```diff
@@ -0,0 +1,232 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.prestosql.sql.planner.iterative.rule;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.Range;
+import io.prestosql.matching.Capture;
+import io.prestosql.matching.Captures;
+import io.prestosql.matching.Pattern;
+import io.prestosql.spi.block.SortOrder;
+import io.prestosql.sql.planner.OrderingScheme;
+import io.prestosql.sql.planner.Symbol;
+import io.prestosql.sql.planner.iterative.Lookup;
+import io.prestosql.sql.planner.iterative.Rule;
+import io.prestosql.sql.planner.plan.Assignments;
+import io.prestosql.sql.planner.plan.JoinNode;
+import io.prestosql.sql.planner.plan.PlanNode;
+import io.prestosql.sql.planner.plan.ProjectNode;
+import io.prestosql.sql.planner.plan.TopNNode;
+import io.prestosql.sql.tree.Expression;
+import io.prestosql.sql.tree.SymbolReference;
+
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+
+import static io.prestosql.matching.Capture.newCapture;
+import static io.prestosql.sql.planner.optimizations.QueryCardinalityUtil.extractCardinality;
+import static io.prestosql.sql.planner.plan.JoinNode.Type.FULL;
+import static io.prestosql.sql.planner.plan.JoinNode.Type.LEFT;
+import static io.prestosql.sql.planner.plan.JoinNode.Type.RIGHT;
+import static io.prestosql.sql.planner.plan.Patterns.Join.type;
+import static io.prestosql.sql.planner.plan.Patterns.join;
+import static io.prestosql.sql.planner.plan.Patterns.project;
+import static io.prestosql.sql.planner.plan.Patterns.source;
+import static io.prestosql.sql.planner.plan.Patterns.topN;
+import static io.prestosql.sql.planner.plan.TopNNode.Step;
+
+/**
+ * Transforms:
+ * <pre>
+ * - TopN
+ *    - [Project]
+ *       - Join
+ *          - left source
+ *          - right source
+ * </pre>
+ * Into:
+ * <pre>
+ * - TopN
+ *    - [Project]
+ *       - Join
+ *          - TopN (present if Join is left or outer and orderBy symbols come from left source)
+ *             - left source
+ *          - TopN (present if Join is right or outer and orderBy symbols come from right source)
+ *             - right source
+ * </pre>
+ */
+public class PushTopNThroughOuterJoinRuleSet
```

Comment:
Make pushdown through join and projection separate rules.

You can pushdown through projection without join.

Reply:
@sopel39 are you sure that Projection pull-up above TopN is a good idea?

Reply:
Projection cost CPU cycles, but you are right we shouldn't push always.
For instance we shouldn't push `TopN` through projection when there is a filter below (as projection get's merged with filter in execution).
We also shouldn't pushdown `TopN` through projection above table scan.
For other cases I think we can pushdown

Reply:
Patterns do not handle the "is not filter nor table scan", so I think we can end up having push-topn-through-x-y-or-z rules. And what we have here would just be one of them.